lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -170,17 +175,21 @@ startOnSocket settings socket wl = Warp.runSettingsSocket settings socket
     application :: Application
     application = serve (Proxy @("v2" :> Api t)) server
 
--- | Run an action with a TCP socket bound to a port. If no port is specified,
--- then an unused port will be selected at random.
+-- | Run an action with a TCP socket bound to a port. If random port is chosen
+-- then an unused port will be selected at random. If exact port is chosen with
+-- no port specified then 9080 will be used.
 withListeningSocket
-    :: Maybe Warp.Port
+    :: PortOption
     -> ((Warp.Port, Socket) -> IO ())
     -> IO ()
-withListeningSocket mport = bracket acquire release
+withListeningSocket portOpt = bracket acquire release
   where
-    acquire = case mport of
-        Just port -> (port,) <$> bindPortTCP port hostPreference
-        Nothing -> bindRandomPortTCP hostPreference
+    acquire = case portOpt of
+        RandomPort -> bindRandomPortTCP hostPreference
+        ExactPort portM -> case portM of
+            Just port -> (port,) <$> bindPortTCP port hostPreference
+            Nothing ->
+                (\port -> (port,) <$> bindPortTCP port hostPreference) 9080
```

Comment:
I have fixed some of the port number duplication.
The default port comes from the docopt text.
But I don't think magic numbers are too bad in tests.
Also it's often good not to share constants between the app and tests.

Reply:
I accept the argument that sometimes it's not good to share constants between the app and tests.

However, if a constant that appears in a test is hard-coded many times within the test suite, and if every instance of that constant is supposed to be the same, then I still think it would be better to use a named constant (that is not shared with the app). Otherwise, readers of the code are left wondering where the number came from.